streams: Message users when someone unsubscribes them from stream(s). (#29272)

tnmkr opened the pull request:
These "you were unsubscribed" messages are supposed to mirror "you were subscribed" messages we already send. So, **we dont message when unsubcriptions occur from management commands, archiving streams, etc**. We only message when someone unsubscribes the user from the api endpoint.

I **did not test** bulk unsubscribe as i do not know of a way to do that from the ui, without archiving the stream.

<!-- Describe your pull request here.-->

Fixes: #29272 

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

![image](https://github.com/zulip/zulip/assets/133781250/23faadfb-5692-4774-aa2a-b3abe26a8cfb)
<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

tnmkr committed:
streams: Add a test for removing user from multiple streams.

We had this ability in the api, but was unused and untested.

tnmkr reviewed (commented):
I do not know if these new query counts are good or bad. I read somewhere that sending messages is expensive(?), so i did not put extra effort trying to optimize these.

tnmkr commented on zerver/views/streams.py:
```diff
@@ -518,27 +532,86 @@ def remove_subscriptions_backend(
+                user_full_name=f"@**{acting_user.full_name}|{acting_user.id}**",
+            )
+        else:
+            message = _("{user_full_name} unsubscribed you from the following streams:").format(
```
Need prep commit for this branch to be included in test coverage. The api support this, but i couldn't find how to do this in the web app. The original issue is unclear on whether we need this, but as seen in the diff, this behaviour was accounted for previously.

prakhar1144 replied:
If the API supports this, then handling it is the right decision. -- otherwise no message would be sent in that case.

tnmkr commented on zerver/tests/test_subs.py:
```diff
@@ -2701,7 +2701,7 @@ def test_admin_remove_others_from_stream_legacy_emails(self) -> None:
     def test_admin_remove_multiple_users_from_stream_legacy_emails(self) -> None:
         result = self.attempt_unsubscribe_of_principal(
-            query_count=20,
+            query_count=41,
```
Query breakdown by function
```
get_system_bot: 1
internal_prep_message: 8 (total 9)
do_send_messages: 11 (21)
```

Adds 5 queries for 2 users.

tnmkr commented on zerver/tests/test_subs.py:
```diff
@@ -2687,7 +2687,7 @@ def test_cant_remove_others_from_stream_legacy_emails(self) -> None:
     def test_admin_remove_others_from_stream_legacy_emails(self) -> None:
         result = self.attempt_unsubscribe_of_principal(
-            query_count=17,
+            query_count=33,
```
Query breakdown by function. Same for others that go from `17 -> 33`.
```
get_system_bot: 1
internal_prep_message: 7 (total 8)
do_send_messages: 8 (16)
```

tnmkr commented:
@alya This is ready for review :)

alya commented:
What happens when you're unscubscribed from multiple stream? Do you get multiple messages?

alya commented:
> I did not test bulk unsubscribe as i do not know of a way to do that from the ui, without archiving the stream.

What happens when a stream is archived?

alya commented:
These messages should use [silent mentions](https://zulip.com/help/mention-a-user-or-group#silently-mention-a-user); please add a commit updating the "subscribed" message as well.

tnmkr committed:
streams: Use silent mention in "subscribed" notification bot messages.

tnmkr commented:
**Updates**
- Added commit to use silent mention syntax in existing messages, as requested above.

---

> What happens when you're unsubscribed from multiple streams? Do you get multiple messages?

No you get a single message with a sorted list of streams. This is similar to the message for subscribing to multiple streams. Though, i don't think there is a way to do either from the ui, here's what that would look like:
![image](https://github.com/zulip/zulip/assets/133781250/f03b7abf-1e74-4445-880f-dde857bba91c)

> What happens when a stream is archived?

We don't send any unsubscription messages in this case.

alya commented:
Looks good to me in manual testing (for subscription changes one stream at a time)!

alya commented:
@prakhar1144 perhaps you could review this one?

tnmkr commented:
Hi @prakhar1144, gentle reminder that this has been sitting for a while :smiley:  

If you can suggest some areas where the query count can be optimized, I would be happy to look into those!

prakhar1144 commented on zerver/views/streams.py:
```diff
@@ -518,27 +532,86 @@ def remove_subscriptions_backend(
+    if len(removed_subscribers) == 0:
+        return
+
+    user_id_to_unsubscribed_streams: Dict[Any, Set[str]] = {}
```
Why are we using `Any` type for user_id ? -- We should use `int`.

tnmkr replied:
Sorry, this was wrong. I have changed the type to `int`.

prakhar1144 commented on zerver/views/streams.py:
```diff
@@ -518,27 +532,86 @@ def remove_subscriptions_backend(
+    stream_names: Set[str],
+    *,
+    add_subscription: bool = False,
+    remove_subscription: bool = False,
```
I think introducing `you_were_just_unsubscribed_message` function would be a better choice -- We can avoid the bunch of if..else statements. The `add_subscription` and `remove_subscription` is mutually exclusive, right? So, we don't need to use two arguments here in any case.

tnmkr replied:
Yes, you are right. I have refactored this as suggested.

prakhar1144 commented on zerver/views/streams.py:
```diff
@@ -518,27 +532,86 @@ def remove_subscriptions_backend(
+            user_id_to_user[user.id] = user
+
+    mention_backend = MentionBackend(acting_user.realm_id)
+    all_messages = []
```
Let's call it `message_requests` -- this is how we name messages which are yet to be sent.

tnmkr replied:
Renamed variable.

prakhar1144 commented on zerver/views/streams.py:
```diff
@@ -518,27 +532,86 @@ def remove_subscriptions_backend(
+            )
+        )
+
+    do_send_messages(all_messages)
```
`do_send_messages(message_requests, mark_as_read=[acting_user.id]))`

tnmkr replied:
Changed this call as suggested.

tnmkr committed:
streams: Message users when someone unsubscribes them from streams.

Send "you were unsubscribed" message to the affected user when someone
unsubscribes them from streams. Since these messages are supposed to
mirror "you were subscribed" messages we already send, we dont send
these when unsubcriptions occur from management commands,
archiving streams, etc. We only message when someone else unsubscribes
the user using the `DELETE /json/users/me/subscriptions` api endpoint.

Fixes #29272.

tnmkr commented:
Made changes as suggested in review.

prakhar1144 commented on zerver/views/streams.py:
```diff
@@ -525,20 +539,88 @@ def you_were_just_subscribed_message(
+        if user_id_to_user[user_id].is_bot:
+            continue
+
+        sender = get_system_bot(settings.NOTIFICATION_BOT, user_id_to_user[user_id].realm_id)
```
We can avoid fetching sender in loop to reduce db query. The sender would be the same here, right?

tnmkr replied:
Yup! Our caller ensures that all users are in the same realm. Made the following change, though this did not reduce query count, it did reduce the cache count (as expected) in `test_subs.py`.

```diff
--- a/zerver/views/streams.py
+++ b/zerver/views/streams.py
@@ -582,6 +582,7 @@ def send_messages_for_removed_subscribers(
             user_id_to_user[user.id] = user

     mention_backend = MentionBackend(acting_user.realm_id)
+    sender: Optional[UserProfile] = None
     message_requests = []
     for user_id, stream_names in user_id_to_unsubscribed_streams.items():
         # Our caller already filters deactivated users. We need to check for
@@ -591,7 +592,12 @@ def send_messages_for_removed_subscribers(
         if user_id_to_user[user_id].is_bot:
             continue

-        sender = get_system_bot(settings.NOTIFICATION_BOT, user_id_to_user[user_id].realm_id)
+        # Our caller ensures that acting_user and all users that they unsubscribe are in
+        # the same realm. This allows us to only fetch this bot once and avoid having to
+        # fetch again with unsubscribed user's realm_id on every iteration of the loop.
+        if sender is None:
+            sender = get_system_bot(settings.NOTIFICATION_BOT, acting_user.realm_id)
+
         message = you_were_just_unsubscribed_message(
             acting_user=acting_user,
             recipient_user=user_id_to_user[user_id],
```

timabbott commented on zerver/tests/test_subs.py:
```diff
@@ -2851,6 +2851,36 @@ def test_remove_invalid_user(self) -> None:
+        if result.status_code not in [400]:
+            for name in stream_names:
+                users_subbed_to_streams = self.users_subscribed_to_stream(name, subbed_user.realm)
+            self.assertNotIn(subbed_user, users_subbed_to_streams)
```
This sort of conditional is a very odd pattern for a test -- we expect the status code to be 200, so we should do an `assert_json_success` to verify that, and then not need this conditional.

timabbott commented on zerver/tests/test_subs.py:
```diff
@@ -2626,8 +2626,8 @@ def test_realm_admin_remove_multiple_users_from_stream(self) -> None:
         result = self.attempt_unsubscribe_of_principal(
-            query_count=28,
-            cache_count=8,
+            query_count=51,
```
This is a very large increase in cost; we should look at whether there's a better way to do this.

timabbott replied:
I note there's a `TODO` comment near the top of this function discussing the concern that this is already too many queries -- that's out of date, but it also suggests that we may be able to find optimizations in the existing code.

Can you look at what the precise parts of `internal_prep_message` and `do_send_messages` are that result in queries here? 

I suspect there are likely some easy optimization opportunities, which would be their own commits, but could be significant here -- e.g. checks that do work that isn't necessary for a message sent by a system bot.

timabbott commented on zerver/views/streams.py:
```diff
@@ -529,6 +543,80 @@ def you_were_just_subscribed_message(
+            user_id_to_unsubscribed_streams[user.id].add(stream.name)
+        else:
+            user_id_to_unsubscribed_streams[user.id] = {stream.name}
+            user_id_to_user[user.id] = user
```
You can simplify this code using `defaultdict(set)` for as the initial value for `user_id_to_unsubscribed_streams`.

timabbott commented on zerver/views/streams.py:
```diff
@@ -529,6 +543,80 @@ def you_were_just_subscribed_message(
     return message
 
 
+def you_were_just_unsubscribed_message(
```
`get_just_unsubscribed_message_content` would be a clearer name for this, since it returns a string (the content), not a Message object.

timabbott commented on zerver/tests/test_subs.py:
```diff
@@ -2874,13 +2874,75 @@ def test_remove_user_from_multiple_streams(self) -> None:
+        *#**hümbüǵ**
+        """
+
+        GetSubscribersTest.assert_user_got_bot_message(GetSubscribersTest(), subbed_user, msg)
```
I think rather than adding these two new tests, you should just modify the existing tests that remove users from streams to assert these notifications were sent and verify their content. Since it's an unconditional part of removing a subscriber, at least for now, there's no reason to repeat the test setup to verify the behavior.

tnmkr commented:
Sorry, accidentally re-requested review here. Please ignore, i have not implement suggested changes yet.

tnmkr commented:
Cross linking #16980 as that is also dealing with increased query count due to bulk private messages. 

https://chat.zulip.org/#narrow/stream/92-learning/topic/Stop.20unsubscribing/near/1132885 also has some relevant discussion.

tnmkr committed:
do_send_messages: Add parameter is_channel_unsubscription_notification.

This new parameter, if true, skips calls to
sew_submessages_and_reactions_to_msgs when converting the message to a
cachable dict in MessageDict.wide_dict.

We do not create our channel unsubscription messages with any widgets or
reactions, so we can be sure they will be empty anyway. Assigning these
dict keys preemptively saves us db queries.

This is a prep commit for #29272.

tnmkr committed:
message_send: Add new parameter user_profile_cache.

This new cache for check_message and do_send_messages helps us save db
queries. We are only using it for some functions in channel
unsubscription notification flow.

This is a prep commit for #29272.

tnmkr committed:
do_send_messages: Fetch in bulk and cache UserPresence objects.

This allows us to avoid queries in loop and save db queries. We are
currently only caching for channel unsubscription notification message
flow.

This is a prep commit for #29272.

tnmkr commented:
Updated this PR with some optimizations to reduce db queries. Here are function calls that included DB queries, most of those were being called in loops. I've replace UserProfile fetches with a dict of cached user profile, added a bulk fetch query for UserPresence objects and skipped fetching submessage and reaction fields. The result is a constant 10 increase in query count as opposed to increases ranging from 16 to 23 for the tests. 

TODOs:-
- Address review comments
- Complete optimizing db queries, there are some reamining UserProfile queries that I can reaplace with the cache.
- Manually test unsubscribing larger number of users. Maybe use populate_db here?
- Update help center documentation as request in issue.
